Your job is to analyse the dataset and perform two tasks

• 1) Find the answers to a set of analytical questions related to the dataset.
Specify how you obtained those answers as well.

• 2) Visualize the required information regarding the dataset

In [1]:
%config IPCompleter.greedy=True

In [78]:
import pandas as pd
import datetime as dt
import numpy as np
from scipy.spatial import distance

# Task 1: Data Analytics Questions: 

## 1. How many active cases (infected individuals) were there on 2020-03-01 (01 March 2020)?

In [3]:
# Read data from file 'df_0000001_20200922055707.csv' 
# (in the same directory that your python process is based)
# Control delimiters, rows, column names with read_csv (see later)
dataset = pd.read_csv("df_0000001_20200922055707.csv")

c:\users\ahsan ijaz\appdata\local\programs\python\python39\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (7,8,9,10,11,12,39) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [4]:
dataset1 = dataset.copy()

In [5]:
# if (data.empty):
#     print("data is loaded")

# Preview the first 5 lines of the loaded data 
dataset.head()
# data.tail()

Unnamed: 0            Date_Time     id  age  gender infection_status  \
0           0  2020-01-01 00:00:00  32645   29       1      susceptible   
1           1  2020-01-01 00:00:00  20022   30       1      susceptible   
2           2  2020-01-01 00:00:00  18409   32       0      susceptible   
3           3  2020-01-01 00:00:00  34113   29       1      susceptible   
4           4  2020-01-01 00:00:00  40168   58       1      susceptible   

   days_since_infection date_of_infection date_of_symptoms date_of_recovery  \
0                   0.0               NaN              NaN              NaN   
1                   0.0               NaN              NaN              NaN   
2                   0.0               NaN              NaN              NaN   
3                   0.0               NaN              NaN              NaN   
4                   0.0               NaN              NaN              NaN   

   ... infected_from_user_id transmitted_count asymptomatic_transmission  \
0  ...                    -1                 0                        -1   
1  ...                    -1                 0                        -1   
2  ...                    -1                 0                        -1   
3  ...                    -1                 0                        -1   
4  ...                    -1                 0                        -1   

   locationID_of_infection        income  household_size  bmi date_of_test  \
0                       -1   5774.248719               6   25          NaN   
1                       -1  14384.879182               1   24          NaN   
2                       -1   9502.567895               4   24          NaN   
3                       -1   7029.974753               5   20          NaN   
4                       -1  12378.711455               2   31          NaN   

   test_result                Scenario  
0           -1  0000001_20200922055707  
1           -1  0000001_20200922055707  
2           -1  0000001_20200922055707  
3           -1  0000001_20200922055707  
4           -1  0000001_20200922055707  

[5 rows x 42 columns]

In [6]:
task1_data = dataset
task1_data["Date"] = pd.to_datetime(task1_data["Date_Time"]).dt.date #Extracting date in a new column
task1_data["Time"] = pd.to_datetime(task1_data["Date_Time"]).dt.time #Extracting time in a new column
task1_data["Date_Time"] = pd.to_datetime(task1_data["Date_Time"])
# Define dates as datetime objects
Given_date = dt.datetime(2020,3,1)
Given_date = pd.to_datetime(Given_date)

temp_data = task1_data.copy() 
task1_data = task1_data[(task1_data['Date'] == Given_date) & (task1_data['infection_status'] == 'infected')]

# task2_data.to_csv('./task2.csv')
task1_data.id.nunique()

# Hamna sy pochna hy
# df_3['movie_title'] = df_3['movie_title'].str.extract('\b([^\d\W]+)\b')



1031

ANSWER: There were 1031 Active Cases on 2020-03-01

## 2.How many people died on 2020-03-01 (01 March 2020)? 

In [7]:
task2_data = temp_data.copy()
task2_data = task2_data[(task2_data['Date'] == Given_date) &(task2_data['infection_status'] == 'deceased')]
# print(task2_data)
task2_data.id.nunique()
# id_dates = dataset[(dataset["id"] == 37415)


49

ANSWER: 49 People were dide on 2020-03-01


## 3.What is the total number of infections, recoveries and deaths? 

In [8]:
task3_data = dataset.copy()

In [9]:
task31_data = task3_data[(task3_data['infection_status'] == 'infected')]
print(task31_data.id.nunique())

task32_data = task3_data[(task3_data['infection_status'] == 'recovered')]
print(task32_data.id.nunique())

task33_data = task3_data[(task3_data['infection_status'] == 'deceased')]
print(task33_data.id.nunique())

# task34_data = task3_data[(task3_data['infection_status'] == 'susceptible')]
# print(task34_data.id.nunique())


3854
3658
215


## 4.How many unique individuals in the dataset?

In [10]:
task4_data = dataset.id.nunique()
print(task4_data)

10000


## 5.How many unique locations and their names in the dataset?

In [19]:
task5_data = dataset.currentLocationID.nunique()
print(task5_data)
task51_data = dataset.copy()
task51_data = task51_data.loc[:,('currentLocationID','currentLocationType')].drop_duplicates()
task51_data


4575


currentLocationID currentLocationType
0                      282         residential
1                     2247         residential
2                     2122         residential
3                      422         residential
4                     1647         residential
...                    ...                 ...
6664706                104         residential
6710969               1447         residential
6744934               2829         residential
6765332               2077         residential
6790486                867         residential

[4575 rows x 2 columns]

## 6. Which school is the closest to a residence ID = 3115? Hint: Use Euclidean distance using the lat and long.

In [133]:
points = []

task6_data_3115 = dataset[(dataset['residence_id'] == 3115)]
task6_data_3115[['residence_id','school_id','lat','lon','currentLocationID','currentLocationType']].head(5)
lat_3115 = task6_data_3115.lat.values[0]
lon_3115 = task6_data_3115.lon.values[0]

task6_data_school = dataset[(dataset['currentLocationType'] == 'school')].drop_duplicates(subset = ['currentLocationID'])
task6_data_school[['residence_id','school_id','lat','lon','currentLocationID','currentLocationType']]

pass_to_loop = task6_data_school[['lat','lon','school_id']]
fixed_point = (lat_3115,lon_3115)


for index, row in pass_to_loop.iterrows():
#     print(row['lat'], row['lon'])
    change_point = (row['lat'],row['lon'])
    points.append([int(row['school_id']),row['lat'],row['lon'],distance.euclidean(fixed_point,change_point)])
loop_out = pd.DataFrame(points, columns=["school_id", "latitude","longitude","closest_distance"])
loop_out[(loop_out['closest_distance'] == loop_out.closest_distance.min())]




# print(type(fixed_point))
# d = distance.euclidean(p1, p2)

school_id  latitude   longitude  closest_distance
35       3563 -6.204529  106.797409          0.014381

## 7. What percentage of individuals aged b/w 30 and 60 visit schools at 08:00 am?

In [251]:
Given_Time = '08:00:00'
Given_Time = dt.datetime.strptime(Given_Time, '%H:%M:%S')
Given_Time = Given_Time.time()
# print('Time:', Given_Time)
# print(type(Given_Time))
# print(type(dataset.Time[0]))

task7_data = dataset[(dataset['Time'] == Given_Time) & (dataset['currentLocationType'] == 'school')]
task7_data = task7_data.drop_duplicates(subset=['id'])

denominator = task7_data[['id']].nunique()
task7_data[['age']]
task7_age = task7_data[(task7_data['age'] >= 30) & (task7_data['age'] <= 60)]
numerator = task7_age[['age']].count()
print(float(numerator))
print(float(denominator))
result = float(numerator)*100/float(denominator)
result



4816.0
6233.0


77.26616396598749

## 8. What is the most dominant daily path of individual (ID= 37415) ? Hint: There are 5 timestamps per day (00:00,08:00,16:00,19:00,22:00). An individual can be at his residence (currentLocationType) at given timestamp, then can move to school on the next time stamp and so on. COMPLEX QUESTION! Can skip it if too difficult.

In [207]:
dataset.currentLocationType.unique()

array(['residential', 'shopping_mall', 'market', 'school', 'employment',
       'hospital'], dtype=object)

In [334]:
Given_Time_00 = '00:00:00'
Given_Time_00 = dt.datetime.strptime(Given_Time_00, '%H:%M:%S')
Given_Time_00 = Given_Time_00.time()

Given_Time_08 = '08:00:00'
Given_Time_08 = dt.datetime.strptime(Given_Time_08, '%H:%M:%S')
Given_Time_08 = Given_Time_08.time()

Given_Time_16 = '16:00:00'
Given_Time_16 = dt.datetime.strptime(Given_Time_16, '%H:%M:%S')
Given_Time_16 = Given_Time_16.time()

Given_Time_19 = '19:00:00'
Given_Time_19 = dt.datetime.strptime(Given_Time_19, '%H:%M:%S')
Given_Time_19 = Given_Time_19.time()

Given_Time_22 = '22:00:00'
Given_Time_22 = dt.datetime.strptime(Given_Time_22, '%H:%M:%S')
Given_Time_22 = Given_Time_22.time()

task8_data = dataset[dataset['id'] == 37415]
# task8_data[['Time','currentLocationID','currentLocationType']]

task8_time_00 = task8_data[task8_data['Time'] == Given_Time_00]
task8_time_00 = task8_time_00['currentLocationType']
# task8_time_00 = task8_time_00.to_frame()
# task8_time_00 = task8_time_00.rename(columns={'currentLocationType': 'Given_Time_00'})
print(task8_time_00)

task8_time_08 = task8_data[task8_data['Time'] == Given_Time_08]
task8_time_08 = task8_time_08['currentLocationType']
# task8_time_08 = task8_time_08.to_frame()
# task8_time_08 = task8_time_08.rename(columns={'currentLocationType': 'Given_Time_08'})
print(task8_time_08)

task8_time_16 = task8_data[task8_data['Time'] == Given_Time_16]
task8_time_16 = task8_time_16['currentLocationType']
# task8_time_16 = task8_time_16.to_frame()
# task8_time_16 = task8_time_16.rename(columns={'currentLocationType': 'Given_Time_16'})
# print(task8_time_16)

task8_time_19 = task8_data[task8_data['Time'] == Given_Time_19]
task8_time_19 = task8_time_19['currentLocationType']
# task8_time_19 = task8_time_19.to_frame()
# task8_time_19 = task8_time_19.rename(columns={'currentLocationType': 'Given_Time_19'})
# print(task8_time_19)

task8_time_22 = task8_data[task8_data['Time'] == Given_Time_22]
task8_time_22 = task8_time_22['currentLocationType']
# task8_time_22 = task8_time_22.to_frame()
# task8_time_22 = task8_time_22.rename(columns={'currentLocationType': 'Given_Time_22'})
# print(task8_time_22)

# task8_time_08 = task8_time_08["Given_Time_08"]

# task8_time_00 = task8_time_00.join(task8_time_08)

# task8_time_00['Given_Time_08'] = task8_time_08


# task8_time_00

hassan = []
hassan.append(task8_time_00)
hassan.append(task8_time_08)
hassan.append(task8_time_16)
hassan.append(task8_time_19)
hassan.append(task8_time_22)
# print(hassan)
hassan_np = np.array(hassan)
transpose = hassan_np

transpose_list = transpose.tolist()

df = pd.DataFrame(transpose_list)
df = df.transpose()
df.columns = ['a', 'b','c','d','e']
af = df.copy()

df.pivot_table(index=['a', 'b','c','d','e'], aggfunc='size')


# af['mix'] = af['0']+af['1']+af['2']+af['3']+af['4']
# freq = np.count_nonzero(path==hassan[0])

# task8_time_08['Given_Time_08']
# task8_data['Given_Time_00'] = task8_time_00['currentLocationType']

159       residential
2319      residential
4479           market
6639      residential
8799      residential
             ...     
283119    residential
285279    residential
287439    residential
289599    residential
291759    residential
Name: currentLocationType, Length: 136, dtype: object
591        employment
2751       employment
4911      residential
7071       employment
9231           school
             ...     
283551    residential
285711     employment
287871     employment
290031     employment
292191     employment
Name: currentLocationType, Length: 136, dtype: object


a              b              c              d              e            
hospital       hospital       employment     residential    residential       1
                              hospital       hospital       hospital         20
market         residential    employment     residential    residential       1
               school         employment     residential    residential       1
residential    employment     employment     employment     residential       5
                                             market         residential       8
                                                            shopping_mall     2
                                             residential    market            6
                                                            residential      24
                                                            shopping_mall     1
                                             shopping_mall  residential       6
                              market         m

## 9. What is the proximity (number of individuals sharing the same location) of an individual ID (37415) at a given date and time? 

In [389]:
Given_Time = '00:00:00'
Given_Time = dt.datetime.strptime(Given_Time, '%H:%M:%S')
Given_Time = Given_Time.time()
Given_date = dt.datetime(2020,3,1)
Given_date = pd.to_datetime(Given_date)

task9_data = dataset[dataset['id'] == 37415]
task9_time = task9_data[(task9_data['Time'] == Given_Time) & (task9_data['Date'] == Given_date)]
location = task9_time['currentLocationID']
location = int(location)
# print(type(location))
task9_2 = dataset[dataset['currentLocationID'] == location]

task9_2['id'].count()



1301

## 10. What is the age and gender of individual 37415? Did this individual get infected? If yes, how long did the infection last? And did the individual die?


In [442]:
task_10 = dataset[(dataset['id'] == 37415)]
age = task_10['age'].unique()
print("Age is :",int(age))
gender = task_10['gender'].unique()
int(gender)
if(gender == 1):
    print("Gender is : Male")
else:
    print("Gender is : FeMale")
    
infected = task_10[(task_10['infection_status'] == 'infected')]
infected_status = infected['infection_status'].unique()
# infected.to_string()
infected_status = np.array_str(infected_status)
print("Yes the individual get ",infected_status)

df_infected = infected[infected.infection_status == 'infected'].drop_duplicates(subset = 'Date')
Days_infected = df_infected.Date.size
int(Days_infected)
print("Number of Days the individual get ",Days_infected)

# dt_infection =infected['date_of_infection']
# dt_recovery = infected['date_of_recovery']
dt_death = infected['date_of_death']

if (dt_death.isnull().values.all() == True):
    print("No the Individual did not die")

Age is : 30
Gender is : Male
Yes the individual get  ['infected']
Number of Days the individual get  29
No the Individual did not die


## 11. What is the infection and mortality rate of health care workers?

In [448]:
dataset[(dataset['type_of_job'] == 'Health worker')]

Unnamed: 0           Date_Time     id  age  gender infection_status  \
1                 1 2020-01-01 00:00:00  20022   30       1      susceptible   
9                 9 2020-01-01 00:00:00  36967   26       1      susceptible   
36               36 2020-01-01 00:00:00  37682   32       1      susceptible   
39               39 2020-01-01 00:00:00  20205   34       1      susceptible   
76               76 2020-01-01 00:00:00  17963   44       0      susceptible   
...             ...                 ...    ...  ...     ...              ...   
6799985     6799985 2020-05-15 00:00:00  32424   42       0        recovered   
6799988     6799988 2020-05-15 08:00:00  32424   42       0        recovered   
6799991     6799991 2020-05-15 16:00:00  32424   42       0        recovered   
6799994     6799994 2020-05-15 19:00:00  32424   42       0        recovered   
6799997     6799997 2020-05-15 22:00:00  32424   42       0        recovered   

         days_since_infection    date_of_infection date_of_symptoms  \
1                         0.0                  NaN              NaN   
9                         0.0                  NaN              NaN   
36                        0.0                  NaN              NaN   
39                        0.0                  NaN              NaN   
76                        0.0                  NaN              NaN   
...                       ...                  ...              ...   
6799985                   0.0  2020-02-19 03:00:00              NaN   
6799988                   0.0  2020-02-19 03:00:00              NaN   
6799991                   0.0  2020-02-19 03:00:00              NaN   
6799994                   0.0  2020-02-19 03:00:00              NaN   
6799997                   0.0  2020-02-19 03:00:00              NaN   

            date_of_recovery  ... asymptomatic_transmission  \
1                        NaN  ...                        -1   
9                        NaN  ...                        -1   
36                       NaN  ...                        -1   
39                       NaN  ...                        -1   
76                       NaN  ...                        -1   
...                      ...  ...                       ...   
6799985  2020-02-21 22:00:00  ...                         1   
6799988  2020-02-21 22:00:00  ...                         1   
6799991  2020-02-21 22:00:00  ...                         1   
6799994  2020-02-21 22:00:00  ...                         1   
6799997  2020-02-21 22:00:00  ...                         1   

        locationID_of_infection        income  household_size  bmi  \
1                            -1  14384.879182               1   24   
9                            -1   9827.205577               4   19   
36                           -1   9554.584788               3   19   
39                           -1   4305.499006               7   23   
76                           -1  12480.380143               1   31   
...                         ...           ...             ...  ...   
6799985                     987   2739.946032               8   25   
6799988                     987   2739.946032               8   25   
6799991                     987   2739.946032               8   25   
6799994                     987   2739.946032               8   25   
6799997                     987   2739.946032               8   25   

         date_of_test  test_result                Scenario        Date  \
1                 NaN           -1  0000001_20200922055707  2020-01-01   
9                 NaN           -1  0000001_20200922055707  2020-01-01   
36                NaN           -1  0000001_20200922055707  2020-01-01   
39                NaN           -1  0000001_20200922055707  2020-01-01   
76                NaN           -1  0000001_20200922055707  2020-01-01   
...               ...          ...                     ...         ...   
6799985           NaN           -1  0000001_20200922055707  2020-05-15   
67999

In [ ]:
plt.plot(dates_rec, recoveries_list, color='red', label = 'Recoveries')
plt.title("Cases W.R.T Dates")
plt.legend()
plt.xlabel("Dates")
plt.ylabel("Cases")
plt.xticks(dates_inf[::29])
plt.show()
# From Muhammad Hassan Maqsood to Everyone:  10:03 PM
df[(df.infection_status == “infected”)].Date.unique()
